### ISSUE TITLE
BetterTransformer slower than normal pytorch

### ISSUE THREAD

**Issue by nreimers** [huggingface/optimum#626]
### System Info

```shell
optimum==1.5.2
torch==1.13.1
transformers==4.25.1
cuda 11.6
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I was hoping to get some inference speedup using BetterTransformer following the HF / pytorch blog post when running larger bulk embedding jobs (e.g. encoding 1M docs with 256 sequence length).

But in fact it is slightly slower than normal Pytorch inference on a DGX machine with V100 GPU.

Here the code:
```python
from transformers import AutoModel, AutoTokenizer
from optimum.bettertransformer import BetterTransformer
import torch
from argparse import ArgumentParser
import tqdm
import time

def benchmark_model(model, encoded):
    #Warmup
    for _ in range(10):
        ret = model(**encoded)

    start_time = time.time()
    with torch.inference_mode():
        for _ in tqdm.trange(args.docs):
            ret = model(**encoded)
    total_time = time.time() - start_time
    total_docs = args.batch_size * args.docs
    print("Docs/Second: {:.2f}".format(total_docs/total_time))


def main(args):
    device="cuda"

    model_name = "bert-base-uncased"
    tokenizer = AutoTokenizer.from_pretrained(model_name)

    doc = "This is a doc "*args.max_seq_length
    docs = [doc]*args.batch_size

    encoded = tokenizer(docs, truncation=True, padding=True, return_tensors="pt", max_length=args.max_seq_length).to(device)
    print("Input IDs:", encoded['input_ids'].shape)

    # Normal transformer model
    normal_model = AutoModel.from_pretrained(model_name).to(device).half()

    print("Normal HF model")
    benchmark_model(normal_model, encoded)


    #BetterTransformers
    bt_model = BetterTransformer.transform(normal_model, keep_original_model=False)
    print(bt_model)

    print("BetterTransformer model")
    benchmark_model(bt_model, encoded)
    


if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument("--batch_size", type=int, default=128)
    parser.add_argument("--model", default='bert-base-uncased')
    parser.add_argument("--docs", type=int, default=200) #Number of batches
    parser.add_argument("--max_seq_length", type=int, default=256)
    args = parser.parse_args()

    print(args)
    main(args)
   ```
   
I tested both, with float32 and float16:
```
Float 32
Normal HF model: 260 docs/sec
BetterTransformer:  260 docs/sec

Float 16:
Normal HF model: 1060 docs/sec
BetterTransformer: 1020 docs/sec   
```

Also tested it on a A100-40GB GPU at Google Cloud:
```
Float 16:
Normal HF model: 2060 docs/sec
BetterTransformer: 1490 docs/sec   
```

Is this expected for this case that you get zero inference speedup? 

### Expected behavior

Based on the blog posts, I would have expected at least some speedup.

**Comment by fxmarty** [comment]
Hi @nreimers , thanks for the report. I don't really think it's a bug. The samples your provide introduce no padding. Hence there are no gains from the nested tensors of bettertransformer. cc @younesbelkada 

Where bettertransformer shines IMO is when doing batched processing, where the input length are unpredictable and hence some padding is needed. For reference, here is the code used for benchmark: https://github.com/huggingface/optimum/blob/main/tests/benchmark/benchmark_bettertransformer.py

Here is a demo I did in the "real" world: https://huggingface.co/spaces/fxmarty/bettertransformer-demo

Could you try generating the inputs as:
```python
    docs = []
    for i in range(args.batch_size):
        n_times = np.random.randint(args.max_seq_length // 2, args.max_seq_length)
        doc = "a " * n_times
        docs.append(doc)
```

or better:

```python
    inp_tokens = torch.randint(tokenizer.vocab_size - 1, (args.batch_size, args.max_seq_length)) + 1
    attention_mask = torch.ones((args.batch_size, args.max_seq_length), dtype=torch.int64)
    for i in range(args.batch_size):
        length = np.random.randint(1, args.max_seq_length // 4)

        inp_tokens[i, -length:] = 0
        inp_tokens[i, 0] = 101
        inp_tokens[i, -length - 1] = 102

        attention_mask[i, -length:] = 0

    encoded = {
        "input_ids": inp_tokens.to(device),
        "attention_mask": attention_mask.to(device)
    }
```

**Comment by fxmarty** [comment]
Actually, I think it's a bug.

With this script:

<details>
  <summary>script</summary>

```python
import argparse

import torch
from transformers import AutoModel

from optimum.bettertransformer import BetterTransformer


def get_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--num-batches",
        type=int,
        default=50,
        help="",
    )
    parser.add_argument(
        "--batch-size",
        type=int,
        default=64,
        help="",
    )
    parser.add_argument(
        "--max-seqlen",
        type=int,
        default=256,
        help="",
    )
    parser.add_argument(
        "--model-name",
        type=str,
        default="bert-base-uncased",
        help="",
    )
    parser.add_argument(
        "--use-half",
        action="store_true",
    )
    return parser


def get_batch(batch_size, max_sequence_length, vocab_size=30522):
    tokens = torch.randint(vocab_size - 1, (batch_size, max_sequence_length), dtype=torch.int64) + 1
    mask = torch.ones(batch_size, max_sequence_length, dtype=torch.int64)
    return tokens, mask


def timing_cuda(model, num_batches, input_ids, masks):
    start_event = torch.cuda.Event(enable_timing=True)
    end_event = torch.cuda.Event(enable_timing=True)
    start_event.record()
    for _ in range(num_batches):
        _ = model(input_ids, masks)
    end_event.record()
    torch.cuda.synchronize()
    return (start_event.elapsed_time(end_event) * 1.0e-3) / num_batches


def benchmark(model_name, num_batches, batch_size, max_seqlen, is_half):
    print("Loading model {}".format(model_name))
    hf_model = AutoModel.from_pretrained(model_name, torch_dtype=torch.float16 if is_half else None).eval()
    hf_model = hf_model.to("cuda:0")
    bt_model = BetterTransformer.transform(hf_model, keep_original_model=True)

    input_ids, masks = get_batch(batch_size, max_seqlen)

    input_ids = input_ids.to("cuda:0")
    masks = masks.to("cuda:0")

    # Warmup
    _ = hf_model(input_ids[0].unsqueeze(0), masks[0].unsqueeze(0))
    torch.cuda.synchronize()
    _ = bt_model(input_ids[0].unsqueeze(0), masks[0].unsqueeze(0))
    torch.cuda.synchronize()

    total_hf_time = timing_cuda(hf_model, num_batches, input_ids, masks)
    total_bt_time = timing_cuda(bt_model, num_batches, input_ids, masks)

    return total_bt_time, total_hf_time


if __name__ == "__main__":
    parser = get_parser()
    args = parser.parse_args()

    BATCH_SIZES = [32, 64]
    SEQ_LEN = [64, 128, 256]

    output_file = open("log_{}.csv".format(args.model_name.replace("/", "-")), "w")
    output_file.write(
        "num_batches, batch_size, seq_len, is half, HF time, BT time, Speedup\n"
    )
    for bs in BATCH_SIZES:
        for seq_len in SEQ_LEN:
            max_seqlen = seq_len
            total_bt_time, total_hf_time = benchmark(
                args.model_name,
                args.num_batches,
                bs,
                max_seqlen,
                args.use_half,
            )

            speedup = total_hf_time / total_bt_time

            output_file.write(
                "{},{},{},{},{},{},{}\n".format(
                    args.num_batches,
                    bs,
                    seq_len,
                    args.use_half,
                    total_hf_time,
                    total_bt_time,
                    speedup,
                )
            )
    output_file.close()
```
</details>

I get on an A100:

```
num_batches, batch_size, seq_len, is half, HF time, BT time, Speedup
50,32,64,False,0.028657294921875,0.0321176171875,0.8922609281559112
50,32,128,False,0.050266171875,0.053516000976562506,0.9392736930589082
50,32,256,False,0.108288798828125,0.106308486328125,1.0186279813437256
50,64,64,False,0.0501,0.0580275390625,0.8633831592623383
50,64,128,False,0.102594521484375,0.106371826171875,0.9644896132421695
50,64,256,False,0.20298849609375,0.20241759765625,1.0028203992346036
```

and

```
num_batches, batch_size, seq_len, is half, HF time, BT time, Speedup
50,32,64,True,0.00623931396484375,0.0115255908203125,0.5413443928486245
50,32,128,True,0.009290178833007813,0.012538204345703125,0.740949706741028
50,32,256,True,0.01894197265625,0.019693486328125,0.9618394803564194
50,64,64,True,0.009155419921875,0.01887340576171875,0.48509633276920766
50,64,128,True,0.016542413330078123,0.023968562011718753,0.6901712886234133
50,64,256,True,0.0357490283203125,0.03755935791015625,0.951800837645464
```

**Comment by younesbelkada** [comment]
OK I managed to reproduce the bug and fix it thanks to the guidance of https://github.com/pytorch/pytorch/issues/91305#issuecomment-1363383905 (@drisspg)
The key is to ALWAYS use nested tensors. Appears that we were not using nested tensors in this specific case: https://github.com/huggingface/optimum/blob/d3c282962ecba7a16280eba0770a5479162ce2f4/optimum/bettertransformer/models/encoder_models.py#L213  
With the fix, I get:
`Docs/Second: 3104.12` for BT model and `Docs/Second: 2101.33` for HF model using PyTorch nightlies - on an A100, which is quite nice.
@nreimers can you try to install `optimum` from the patch branch: `pip install git+https://github.com/younesbelkada/optimum@patch-bt` and let us know if this fixes your issue before addressing a proper fix ? Thanks

**Comment by younesbelkada** [comment]
and FYI to install PyTorch nightlies: 
```
pip install https://download.pytorch.org/whl/nightly/cu116/torch-2.0.0.dev20221221%2Bcu116-cp38-cp38-linux_x86_64.whl
```
This is what I have run on the A100 GCP machine under CUDA 11.6

**Comment by fxmarty** [comment]
Hi @NielsRogge , the issue has been addressed on the main branch. We'll do a release today which will include the fix.

You can try [this test](https://github.com/huggingface/optimum/blob/46004525576c8b8aad0e4d6bae737843fc71e49c/tests/bettertransformer/test_gpu.py#L83) to check the fix:

```
TEST_LEVEL=1 pytest tests/bettertransformer/test_gpu.py -s --exitfirst
```

**Comment by regisss** [comment]
@fxmarty I think you wanted to tag @nreimers here :slightly_smiling_face: 
I guess Niels and Nils are pretty used to this from the time Nils was at HF :grin:
